### ISSUE TITLE
実装+実験: exchange compact396入力を導入し2724次元版と比較する

### ISSUE THREAD

**Issue by eletim** [eletim/napoleon-web#438]
Related: #202, #431, #433, #434, #436
Base branch: `feature/issue-202-non-playing-ai`

## 背景

#436 では exchange observation 2671 dims + candidate discard mask 53 dims = 2724 dims を入力とする埋札 combination value modelを学習した。

1,000 unique exchange states / 286,000 samplesまで増やした結果、RuleBasedよりわずかに良い程度でplateauした。

- model margin regret: 4.39
- RuleBased margin regret: 4.59
- pairwise: 0.588
- exact top-action: 0.01
- top5: 0.07

一方、現行 exchange observation 2671 dims は汎用 non-playing encoder由来で、競り履歴を最大長action sequenceとして巨大one-hot展開するなど、ゲームルール上冗長な表現が多い。

競りAIでは compact278 を使っており、これは情報を削るのではなく、ナポレオンの競りの特性

1. より高いbidにしか進まない
2. 4連続PASSで終了する

を利用し、競り履歴を `28 bid positions × (unused + 5 players)` の owner table 等で固定長かつ実質losslessに表現している。

exchangeでも同じ思想で、プレイヤーが知っている情報を落とさず、意味の明確な compact input に再設計する。

## 目的

1. exchange専用 compact state input を正式定義する
2. #436 の同一dataset / 同一split / 同一teacherで compact版を学習する
3. 2724-dim baselineとの比較で、入力表現だけの効果を確認する
4. compact表現でranking/regretが改善するか検証する
5. 改善した場合、今後の埋札・副官AI設計の基礎表現として採用する

## Compact exchange state input

候補discardを除くstate表現は **343 dims** とする。

### 1. 元々の10枚手札: 53 dims

`originalHandMask[53]`

- bidding終了時点、kitty pickup前にNapoleonが持っていた10枚
- 53-card binary mask

### 2. kittyでもらった3枚: 53 dims

`kittyPickupMask[53]`

- 副官指定後にpickupした3枚
- 53-card binary mask
- originalHandMaskとのORで現在の13枚handを完全復元可能

現在13枚hand maskを別途重複して追加しない。

### 3. 副官指定カード: 53 dims

`calledAdjutantCardMask[53]`

- 指定カード1枚のone-hot / mask

### 4. 最終contract suit: 4 dims

`contractSuitOneHot[4]`

- spades/hearts/diamonds/clubs

### 5. 最終contract target: 7 dims

`contractTargetOneHot[7]`

- 13..19

### 6. 競り開始者: 5 dims

`starterPlayerOneHot[5]`

- Napoleon/actor基準relative player indexで表現
- absolute seat IDには依存しない

### 7. 競り履歴: 168 dims

`biddingBidOwnerTableOneHot[28,6]`

- 28 bid positions = target 13..19 × 4 suits
- 各positionのclass:
  - unused
  - relative player 0
  - relative player 1
  - relative player 2
  - relative player 3
  - relative player 4

合計:

```text
53 + 53 + 53 + 4 + 7 + 5 + 168 = 343 dims
```

## Candidate discard input

各286候補を評価するvalue modelなので、candidate unordered 3-card setを追加する。

`candidateDiscardMask[53]`

したがって最終model input:

```text
compact exchange state 343
+ candidate discard mask 53
= 396 dims
```

**compact396** と呼ぶ。

## 情報保持の原則

これは単なるdimension削減ではない。

目標は、exchange時点でNapoleonが持つ関連情報を保ったまま、ゲームルール上冗長なencodingを除くこと。

特に競り履歴は、compact278と同じbid owner table semanticsを再利用する。

### 含めないもの

以下はcompact396では原則不要。

- full bidding chronological action sequence one-hot
- `biddingHistoryActionMask`
- action type / player / suit / target の最大長履歴one-hot
- `partialDiscardMask`（今回のactionはunordered 3-card set一括評価）
- `exchangeStepIndex`
- `remainingDiscardCount`（常に3）
- `legalDiscardCardMask`（original hand + kittyから導出可能）
- `handCountByPlayer`（exchange開始時点でルール上既知/固定）
- special card indices one-hot（card identity + contract suitから導出可能）
- 現在13枚handの重複mask

ただし実装時に、compact396から元の判断に必要な情報が本当に復元/導出可能かコードで監査し、情報欠落がある場合は勝手に削除せずIssueコメントで報告すること。

## 競り履歴のlossless性確認

compact278の考え方をそのまま使う。

ナポレオンの競りでは:

- bid valueは単調に上昇する
- turn orderは固定
- 4連続PASSで終了

したがって、starter + bid owner table + final contractから、交換判断に必要な競り履歴情報が保持されることをtests/fixtureで確認する。

少なくとも既存 #436 dataset の biddingHistory から compact historyへencodeし、同一public auctionに対しdeterministicに同一compact表現になること。

## Dataset

#436で生成済みのmain datasetを原則再利用する。

- 1,000 unique source exchange states
- 286 candidates/state
- 286,000 samples
- train/validation/final = 800/100/100 states

重要:
- teacherを作り直さない
- splitを変えない
- hidden dealを変えない
- candidate actionを変えない

既存datasetにoriginal 10-card hand / kitty 3-card identity等が不足してcompact396を完全再構築できない場合のみ、#434 generator semantics・同seed panelでdatasetを再生成する。
その場合は、#436 baselineとのmatched comparisonが成立するようsourceStateKey / dealSeed / split identitiesを合わせる。

## Baseline

B0 = #436 best model

```text
2724 dims
= old exchange observation 2671
+ candidate mask 53
MLP [512,512,256,256]
Huber regression
```

既存final metrics:
- MAE 2.657
- RMSE 3.310
- Pearson 0.317
- pairwise 0.588
- exact 0.01
- top3 0.03
- top5 0.07
- selected rank percentile 0.675
- margin regret 4.39
- relative reward regret 12.975
- RuleBased margin regret 4.59

## Compact model

C1:

```text
396 -> [512,512,256,256] -> scalar contractMargin
```

まず #436 と同じ:
- hidden dims
- Huber loss delta 1.0
- optimizer
- lr
- batch size
- early stopping
- seed

を使い、**入力表現以外を変えない**。

architecture/loss tuningで差を混ぜない。

## Evaluation

#436と同一final held-out 100 statesで比較する。

### scalar
- MAE
- RMSE
- Pearson
- bias

### same-state ranking
- pairwise accuracy
- exact top-action
- top3 hit
- top5 hit
- selected teacher rank percentile
- selected margin regret
- selected relative reward regret

### RuleBased comparison
- model vs RuleBased win/tie/loss
- mean margin-regret improvement
- mean relative-reward-regret improvement

### bury content
compact model / B0 / RuleBased / teacher-bestについて:
- buried point-card count
- trump count
- special-card bury frequency

異常collapseがないか確認。

## Learning curve

同じsplit/subset定義で可能なら:

- 250 train states
- 500 train states
- 800 train states（1,000 total source states）

を比較する。

B0とcompact396の両方を同じtableに並べる。

## 追加diagnostic: 表現同値性

少なくとも100 source statesで以下を確認する。

- originalHandMask bit count = 10
- kittyPickupMask bit count = 3
- overlap = 0
- OR bit count = 13
- calledAdjutantCardMask bit count = 1
- contract suit one-hot = 1
- contract target one-hot = 1
- starter one-hot = 1
- bid owner table shape = 28×6
- each bid position one-hot = exactly 1
- compact state length = 343
- candidate追加後 = 396

さらに、同一source stateの286 candidatesでstate 343 dimsが完全一致し、candidate 53 dimsだけが変化することを確認する。

### 採用
compact396がB0より明確に改善、または同等性能を大幅に小さい入力で維持した場合、今後exchangeの正式input候補とする。

特に期待するのは:
- pairwise > B0 0.588
- margin regret < B0 4.39
- RuleBased 4.59をより明確に下回る

Exact top-actionはteacher noiseの影響があるため主指標にしない。

### 改善しない場合
compact396でも同様にplateauするなら、次の主因候補はinput dimensionではなく:
- one-hidden-deal teacher noise
- hidden information依存
- expected-value teacher不足

と判断し、同一visible stateに対するhidden-deal N-samplingへ進む。

## 非目標

- runtime wiring
- Frozen artifact化
- 副官NN学習
- joint副官+埋札model
- reward変更
- playing再学習
- bidding AI変更
- hyperparameter探索

## Tests

最低限:
1. compact396 feature count
2. original hand 10-bit mask
3. kitty 3-bit mask
4. original/kitty disjoint + OR=13
5. adjutant 1-hot
6. contract suit/target one-hot
7. relative starter encoding
8. bid owner table 28×6 shape / one-hot
9. compact bidding history parity fixture
10. same source state across 286 candidates shares identical 343 state features
11. candidate mask exactly 3 bits
12. compact dataset loader
13. group split leakage guard unchanged
14. training smoke
15. deterministic final evaluation

## Git / PR

最新 `feature/issue-202-non-playing-ai` から専用branchを作る。

PR base:
`feature/issue-202-non-playing-ai`

main向けPRは禁止。

実装・学習・評価・tests後、Codex review最大4回（review後修正最大3回）。blocking/actionable指摘がなくなればReadyにして親branchへmergeする。

mainにはmergeしない。

## 完了時に必ず報告

- compact396の正確なfeature layout
- 情報保持/lossless性監査結果
- dataset再利用か再生成か
- split identities
- B0 vs compact396 learning curve
- scalar metrics
- pairwise/top-k/regret
- RuleBased比較
- bury content分布
- compact396を正式採用すべきか
- plateau原因をinputから切り離せたか
- 次にexpected-value teacherへ進むべきか
- tests
- PR / merge commit / Codex review回数
- mainへmergeしていないこと
